File: js/189. Rotate Array.js
const reverse = (arr, start, end) => {
    while (start < end) {
        [arr[start], arr[end]] = [arr[end], arr[start]];
        start++;
        end--;
    }
};

var rotate = function(nums, k) {
    const n = nums.length;
    k = k % n;

    if (k === 0) return;

    reverse(nums, 0, n - 1);
    reverse(nums, 0, k - 1);
    reverse(nums, k, n - 1);
};

File: python/leetcode/189. Rotate Array.py
def _reverse(self, nums, l, r):
        while l < r:
            nums[l], nums[r] = nums[r], nums[l]
            l, r = l + 1, r - 1

def rotate(self, nums, k):
    k = k % len(nums)

    self._reverse(nums, 0, len(nums) - 1)
    self._reverse(nums, 0, k - 1)
    self._reverse(nums, k, len(nums) - 1)